mava/networks.py

```diff
@@ -0,0 +1,253 @@
+    post_torso: nn.Module
+
+    @nn.compact
+    def __call__(
+        self,
+        policy_hidden_state: chex.Array,
+        observation_done: RNNObservation,
+    ) -> Tuple[chex.Array, distrax.Categorical]:
+        """Forward pass."""
+        observation, done = observation_done
+
+        policy_embedding = self.pre_torso(observation.agents_view)
+        policy_rnn_in = (policy_embedding, done)
+        policy_hidden_state, policy_embedding = ScannedRNN()(policy_hidden_state, policy_rnn_in)
+        actor_output = self.post_torso(policy_embedding)
+        actor_output = nn.Dense(
+            self.action_dim, kernel_init=orthogonal(0.01), bias_init=constant(0.0)
+        )(actor_output)
+
+        masked_logits = jnp.where(
+            observation.action_mask,
+            actor_output,
+            jnp.finfo(jnp.float32).min,
+        )
+
+        pi = distrax.Categorical(logits=masked_logits)
+
+        return policy_hidden_state, pi
+
+
+class RecurrentCritic(nn.Module):
+    """Recurrent Critic Network."""
+
+    pre_torso: nn.Module
+    post_torso: nn.Module
+    centralised_critic: bool = False
+
+    @nn.compact
+    def __call__(
+        self,
+        critic_hidden_state: Tuple[chex.Array, chex.Array],
+        observation_done: Union[RNNObservation, RNNGlobalObservation],
+    ) -> Tuple[chex.Array, chex.Array]:
+        """Forward pass."""
+        observation, done = observation_done
+
+        if self.centralised_critic:
+            if not isinstance(observation, ObservationGlobalState):
+                raise ValueError("Global state must be provided to the centralised critic.")
+            # Get global state in the case of a centralised critic.
+            observation = observation.global_state
+        else:
+            # Get single agent view in the case of a decentralised critic.
+            observation = observation.agents_view
+
+        critic_embedding = self.pre_torso(observation)
+        critic_rnn_in = (critic_embedding, done)
+        critic_hidden_state, critic_embedding = ScannedRNN()(critic_hidden_state, critic_rnn_in)
+        critic_output = self.post_torso(critic_embedding)
+        critic_output = nn.Dense(1, kernel_init=orthogonal(1.0), bias_init=constant(0.0))(
+            critic_output
+        )
+
+        return critic_hidden_state, jnp.squeeze(critic_output, axis=-1)
+
+
+def get_networks(
+    config: Dict, network: str, centralised_critic: bool = False
+) -> Union[Tuple[FeedForwardActor, FeedForwardCritic], Tuple[RecurrentActor, RecurrentCritic]]:
+    """Get the networks."""
+
+    def create_torso(network_key: str, layer_size_key: str) -> MLPTorso:
+        """Helper function to create a torso object from the config."""
+        return MLPTorso(
+            layer_sizes=config["system"][network_key][layer_size_key],
+            activation=config["system"][network_key]["activation"],
+            use_layer_norm=config["system"][network_key]["use_layer_norm"],
+        )
```

Comment:
Can we add this for parsing activation, keeps it decoupled and adds ability to more easily add new activation functions

```suggestion
def parse_activation_fn(activation_fn_name: str):
    activation_fns = {"relu": nn.relu, "tanh": nn.tanh}
    return activation_fns[activation_fn_name]

def get_networks(
    config: Dict, network: str, centralised_critic: bool = False
) -> Union[Tuple[FeedForwardActor, FeedForwardCritic], Tuple[RecurrentActor, RecurrentCritic]]:
    """Get the networks."""

    def create_torso(network_key: str, layer_size_key: str) -> MLPTorso:
        """Helper function to create a torso object from the config."""
        activation_fn = parse_activation_fn(config["system"][network_key]["activation"])
            
        return MLPTorso(
            layer_sizes=config["system"][network_key][layer_size_key],
            activation=activation_fn,
            use_layer_norm=config["system"][network_key]["use_layer_norm"],
        )
```